File: Sources/custom_adapter/AdmobAdapter/Objc/AlxAdmobSplashAdapter.m
//
//  AlxAdmobSplashAdapter.m
//  AlxAdsOCDemo
//

#import "AlxAdmobSplashAdapter.h"
#import <AlxAds/AlxAds-Swift.h>

static NSString *const TAG = @"AlxAdmobSplashAdapter";

@interface AlxAdmobSplashAdapter () <AlxSplashAdDelegate>

@property (nonatomic, strong, nullable) AlxSplashAd *splashAd;
@property (nonatomic, weak, nullable) id<GADMediationAppOpenAdEventDelegate> delegate;
@property (nonatomic, copy, nullable) GADMediationAppOpenLoadCompletionHandler completionHandler;

@end

@implementation AlxAdmobSplashAdapter

- (void)loadAppOpenAdForAdConfiguration:(GADMediationAppOpenAdConfiguration *)adConfiguration
                      completionHandler:(GADMediationAppOpenLoadCompletionHandler)completionHandler {
    NSLog(@"%@: loadAppOpenAd", TAG);

    NSDictionary *params = [AlxAdmobBaseAdapter parseAdparameterFor:adConfiguration.credentials];
    if (!params) {
        NSString *errorStr = @"The parameter field is not found in the adConfiguration object";
        NSLog(@"%@: config params is empty", TAG);
        self.delegate = completionHandler(nil, [self errorWithCode:-100 msg:errorStr]);
        return;
    }

    if (!AlxAdmobBaseAdapter.isInitialized) {
        [AlxAdmobBaseAdapter initSdkFor:params];
    }

    NSString *adId = params[@"unitid"];
    if (!adId || adId.length == 0) {
        NSString *errorStr = @"unitid is empty in the parameter configuration";
        NSLog(@"%@: error: %@", TAG, errorStr);
        self.delegate = completionHandler(nil, [self errorWithCode:-100 msg:errorStr]);
        return;
    }

    NSLog(@"%@: loadAppOpenAd unitid=%@", TAG, adId);
    self.completionHandler = completionHandler;

    BOOL autoClose = NO;
    if (params[@"autoCloseOnFinish"]) {
        autoClose = [params[@"autoCloseOnFinish"] boolValue];
    } else if (params[@"auto_close"]) {
        autoClose = [params[@"auto_close"] boolValue];
    } else if (params[@"is_auto_close"]) {
        autoClose = [params[@"is_auto_close"] boolValue];
    }

    self.splashAd = [[AlxSplashAd alloc] init];
    self.splashAd.delegate = self;
    self.splashAd.autoCloseOnFinish = autoClose;

    AlxAdRequest *req = [[[AlxAdRequest alloc] init] withUserExt:@{ @"bid_floor": @"1.68" }];
    [self.splashAd loadAdWithAdUnitId:adId request:req];
}

- (void)presentFromViewController:(UIViewController *)viewController {
    NSLog(@"%@: present", TAG);
    if (self.splashAd && [self.splashAd isReady]) {
        [self.splashAd showAdWithPresent:viewController];
    } else {
        NSString *errorStr = @"Splash ad is not ready to present";
        NSLog(@"%@: error: %@", TAG, errorStr);
        [self.delegate didFailToPresentWithError:[self errorWithCode:-101 msg:errorStr]];
    }
}

#pragma mark - AlxSplashAdDelegate

- (void)splashAdDidLoad:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidLoad", TAG);
    if (self.completionHandler) {
        self.delegate = self.completionHandler(self, nil);
    }
}

- (void)splashAdDidFailToLoad:(AlxSplashAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: splashAdDidFailToLoad: %@", TAG, error.localizedDescription);
    if (self.completionHandler) {
        self.delegate = self.completionHandler(nil, error);
    }
}

- (void)splashAdDidShow:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidShow", TAG);
    [self.delegate willPresentFullScreenView];
    [self.delegate reportImpression];
}

- (void)splashAdDidClick:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidClick", TAG);
    [self.delegate reportClick];
}

- (void)splashAdDidClose:(AlxSplashAd *)ad {
    NSLog(@"%@: splashAdDidClose", TAG);
    [self.delegate didDismissFullScreenView];
}

- (void)splashAdRenderDidFail:(AlxSplashAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: splashAdRenderDidFail: %@", TAG, error.localizedDescription);
    [self.delegate didFailToPresentWithError:error];
}

@end

File: Sources/custom_adapter/AdmobAdapter/Swift/AlxAdmobSplashAdapter.swift
//
//  AlxAdmobSplashAdapter.swift
//  AlxAdsDemo
//

import Foundation
import UIKit
import GoogleMobileAds
import AlxAds

@objcMembers
@objc(AlxAdmobSplashAdapter)
public class AlxAdmobSplashAdapter: AlxAdmobBaseAdapter, MediationAppOpenAd {
    
    private static let TAG = "AlxAdmobSplashAdapter"
    
    private var splashAd: AlxSplashAd? = nil
    private var delegate: MediationAppOpenAdEventDelegate? = nil
    private var completionHandler: GADMediationAppOpenLoadCompletionHandler? = nil
    
    public func loadAppOpenAd(for adConfiguration: GADMediationAppOpenAdConfiguration, completionHandler: @escaping GADMediationAppOpenLoadCompletionHandler) {
        NSLog("%@: loadAppOpenAd", AlxAdmobSplashAdapter.TAG)
        guard let params = AlxAdmobBaseAdapter.parseAdparameter(for: adConfiguration.credentials) else {
            let errorStr = "The parameter field is not found in the adConfiguration object"
            NSLog("%@: config params is empty", AlxAdmobSplashAdapter.TAG)
            self.delegate = completionHandler(nil, self.error(code: -100, msg: errorStr))
            return
        }
        
        if !AlxAdmobBaseAdapter.isInitialized {
            AlxAdmobBaseAdapter.initSdk(for: params)
        }
        
        guard let adId = params["unitid"] as? String, !adId.isEmpty else {
            let errorStr = "unitid is empty in the parameter configuration"
            NSLog("%@: error: %@", AlxAdmobSplashAdapter.TAG, errorStr)
            self.delegate = completionHandler(nil, self.error(code: -100, msg: errorStr))
            return
        }
        
        NSLog("%@: loadAppOpenAd unitid=%@", AlxAdmobSplashAdapter.TAG, adId)
        self.completionHandler = completionHandler
        
        // autoCloseOnFinish 配置（默认 false）
        var autoClose: Bool = false
        if let v = params["autoCloseOnFinish"] as? Bool {
            autoClose = v
        } else if let v = params["auto_close"] as? Bool {
            autoClose = v
        } else if let v = params["is_auto_close"] as? Bool {
            autoClose = v
        }
        
        // 开始加载广告
        // Load ad
        self.splashAd = AlxSplashAd()
        self.splashAd?.delegate = self
        self.splashAd?.autoCloseOnFinish = autoClose
        
        // 扩展字段
        let req = AlxAdRequest().withUserExt([
            "bid_floor": "1.68"
        ])
        self.splashAd?.loadAd(adUnitId: adId, request: req)
    }
    
    public func present(from viewController: UIViewController) {
        NSLog("%@: present", AlxAdmobSplashAdapter.TAG)
        DispatchQueue.main.async {
            if let splashAd = self.splashAd, splashAd.isReady() {
                splashAd.showAd(present: viewController)
            } else {
                let errorStr = "Splash ad is not ready to present"
                NSLog("%@: error: %@", AlxAdmobSplashAdapter.TAG, errorStr)
                self.delegate?.didFailToPresentWithError(self.error(code: -101, msg: errorStr))
            }
        }
    }
}

// MARK: - AlxSplashAdDelegate
extension AlxAdmobSplashAdapter: AlxSplashAdDelegate {
    
    public func splashAdDidLoad(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidLoad", AlxAdmobSplashAdapter.TAG)
        if let handler = self.completionHandler {
            self.delegate = handler(self, nil)
        }
    }
    
    public func splashAdDidFailToLoad(_ ad: AlxSplashAd, didFailWithError error: Error) {
        NSLog("%@: splashAdDidFailToLoad: %@", AlxAdmobSplashAdapter.TAG, error.localizedDescription)
        if let handler = self.completionHandler {
            self.delegate = handler(nil, error)
        }
    }
    
    public func splashAdDidShow(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidShow", AlxAdmobSplashAdapter.TAG)
        self.delegate?.willPresentFullScreenView()
        self.delegate?.reportImpression()
    }
    
    public func splashAdDidClick(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidClick", AlxAdmobSplashAdapter.TAG)
        self.delegate?.reportClick()
    }
    
    public func splashAdDidClose(_ ad: AlxSplashAd) {
        NSLog("%@: splashAdDidClose", AlxAdmobSplashAdapter.TAG)
        self.delegate?.didDismissFullScreenView()
    }
    
    public func splashAdRenderDidFail(_ ad: AlxSplashAd, didFailWithError error: Error) {
        NSLog("%@: splashAdRenderDidFail: %@", AlxAdmobSplashAdapter.TAG, error.localizedDescription)
        self.delegate?.didFailToPresentWithError(error)
    }
    
    public func splashAdCountdown(_ ad: AlxSplashAd, countdown: Int) {
        NSLog("%@: splashAdCountdown: %ld", AlxAdmobSplashAdapter.TAG, countdown)
    }
}
